apm-sniffer/apm-toolkit-activation/apm-toolkit-logging-common/src/main/java/org/apache/skywalking/apm/toolkit/logging/common/log/GRPCLogReportServiceClient.java

```diff
@@ -0,0 +1,143 @@
+import org.apache.skywalking.apm.network.logging.v3.LogReportServiceGrpc;
+
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.StatusRuntimeException;
+import io.grpc.stub.StreamObserver;
+
+/**
+ * Report log to server by grpc
+ */
+@OverrideImplementor(LogReportServiceClient.class)
+public class GRPCLogReportServiceClient extends LogReportServiceClient {
+
+    private static final ILog LOGGER = LogManager.getLogger(GRPCLogReportServiceClient.class);
+
+    private volatile DataCarrier<LogData> carrier;
+
+    private LogReportServiceGrpc.LogReportServiceStub asyncStub;
+
+    private ManagedChannel channel;
+
+    private AtomicBoolean disconnected = new AtomicBoolean(false);
+
+    @Override
+    public void boot() throws Throwable {
+        carrier = new DataCarrier<>("gRPC-log", "gRPC-log", Config.Buffer.CHANNEL_SIZE, Config.Buffer.BUFFER_SIZE,
+                BufferStrategy.IF_POSSIBLE);
+        carrier.consume(this, 1);
+        channel = ManagedChannelBuilder
+                .forAddress(ToolkitConfig.Plugin.Toolkit.Log.GRPC.Reporter.SERVER_HOST,
+                        ToolkitConfig.Plugin.Toolkit.Log.GRPC.Reporter.SERVER_PORT).usePlaintext().build();
+        asyncStub = LogReportServiceGrpc.newStub(channel)
+                .withMaxOutboundMessageSize(ToolkitConfig.Plugin.Toolkit.Log.GRPC.Reporter.MAX_MESSAGE_SIZE);
+    }
+
+    @Override
+    public void shutdown() {
+        try {
+            if (channel != null) {
+                channel.shutdownNow();
+            }
+            carrier.shutdownConsumers();
+        } catch (Throwable t) {
+            LOGGER.error(t.getMessage(), t);
+        }
+    }
+
+    @Override
+    public void produce(LogData logData) {
+        if (Objects.nonNull(logData) && !carrier.produce(logData)) {
+            if (LOGGER.isDebugEnable()) {
+                LOGGER.debug("One log has been abandoned, cause by buffer is full.");
+            }
+        }
+    }
+
+    @Override
+    public void consume(final List<LogData> dataList) {
+        if (CollectionUtil.isEmpty(dataList)) {
+            return;
+        }
+        StreamObserver<LogData> reportStreamObserver = null;
+        final GRPCStreamServiceStatus waitStatus = new GRPCStreamServiceStatus(false);
+        try {
+            reportStreamObserver = asyncStub.withDeadlineAfter(
+                    ToolkitConfig.Plugin.Toolkit.Log.GRPC.Reporter.UPSTREAM_TIMEOUT, TimeUnit.SECONDS
+            ).collect(new StreamObserver<Commands>() {
+                @Override
+                public void onNext(Commands commands) {
+                }
+
+                @Override
+                public void onError(Throwable t) {
+                    waitStatus.finished();
+                    if (!disconnected.get()) {
+                        LOGGER.error("Send log to gRPC server fail with an internal exception.", t);
+                        disconnected.set(true);
+                    }
```

Comment:
`disconnected` may be unnecessary, but if it's really needed, it should be

```suggestion
                    if (disconnected.compareAndSet(false, true)) {
                        LOGGER.error("Send log to gRPC server fail with an internal exception.", t);
                    }
```